python/tests/test_tree_stats.py

```diff
@@ -1768,6 +1768,273 @@ class TestSiteDivergence(TestDivergence, MutatedTopologyExamplesMixin):
+############################################
+
+
+def site_genetic_relatedness(
+    ts, sample_sets, indexes, windows=None, span_normalise=True, proportion=True
+):
+    out = np.zeros((len(windows) - 1, len(indexes)))
+    samples = ts.samples()
+    all_samples = list({u for s in sample_sets for u in s})
+    haps = ts.genotype_matrix(isolated_as_missing=False).T
+    denom = np.ones(len(windows))
+    if proportion:
+        denom = ts.segregating_sites(
+            sample_sets=all_samples, windows=windows, mode="site"
+        )
+    alleles = np.unique(haps)
+    for j in range(len(windows) - 1):
+        begin = windows[j]
+        end = windows[j + 1]
+        site_positions = [x.position for x in ts.sites()]
+        for i, (ix, iy) in enumerate(indexes):
+            X = sample_sets[ix]
+            Y = sample_sets[iy]
+            S = 0
+            for a in alleles:
+                this_haps = haps == a
+                haps_mean = this_haps.mean(axis=0)
+                haps_centered = this_haps - haps_mean
+                for k in range(ts.num_sites):
+                    if (site_positions[k] >= begin) and (site_positions[k] < end):
+                        for x in X:
+                            x_index = np.where(samples == x)[0][0]
+                            for y in Y:
+                                y_index = np.where(samples == y)[0][0]
+                                S += (
+                                    haps_centered[x_index][k]
+                                    * haps_centered[y_index][k]
+                                    / 2
+                                )
+            with np.errstate(invalid="ignore", divide="ignore"):
+                out[j][i] = S / denom[j]
+            if span_normalise:
+                out[j][i] /= end - begin
+    return out
+
+
+def branch_genetic_relatedness(
+    ts, sample_sets, indexes, windows=None, span_normalise=True, proportion=True
+):
+    out = np.zeros((len(windows) - 1, len(indexes)))
+    samples = ts.samples()
+    denom = np.ones(len(windows))
+    if proportion:
+        all_samples = list({u for s in sample_sets for u in s})
+        denom = ts.segregating_sites(
+            sample_sets=all_samples, windows=windows, mode="branch"
+        )
+    for j in range(len(windows) - 1):
+        begin = windows[j]
+        end = windows[j + 1]
+        for tr in ts.trees():
+            if tr.interval[1] <= begin:
+                continue
+            if tr.interval[0] >= end:
+                break
+            branches = [(c, tr.parent(c)) for c in tr.nodes()]
+            span = min(end, tr.interval[1]) - max(begin, tr.interval[0])
+            for B in branches:
+                v = B[0]
+                area = np.sqrt(tr.branch_length(v) * span)
+                haps = np.zeros(ts.num_samples)
+                for x, u in enumerate(samples):
+                    haps[x] = np.int(tr.is_descendant(u, v))
+                haps_mean = haps.mean()
```

Comment:
Hm, won't this make the statistic differ depending on what samples are in the tree sequence? Which should not happen? For instance, suppose we have a tree sequence with 10 samples; it should be the case that
```
ts.genetic_relatedness([[0,1], [2,3,4]], mode='branch')
```
is equal to the same thing, computed after removing samples 5-9:
```
sts = ts.simplify([0,1,2,3,4])
sts.genetic_relatedness([[0,1], [2,3,4]], mode='branch')
```
Here's a quick check:
```
ts = msprime.simulate(10, mutation_rate=3, recombination_rate=3, random_seed=23)
branch_genetic_relatedness(ts, [[0,1], [2,3,4]], indexes=[(0,1)], windows=[0, ts.sequence_length])
# array([[0.13483581]])
sts = ts.simplify([0,1,2,3,4])
branch_genetic_relatedness(sts, [[0,1], [2,3,4]], indexes=[(0,1)], windows=[0, ts.sequence_length])
# array([[-0.206612]])
## but all is good for the basic implementation:
ts.genetic_relatedness([[0,1], [2,3,4]], indexes=[(0,1)], windows=[0, ts.sequence_length], mode='branch')
# array([[-0.206612]])
sts.genetic_relatedness([[0,1], [2,3,4]], indexes=[(0,1)], windows=[0, ts.sequence_length], mode='branch')
# array([[-0.206612]])
```

Reply:
I believe that the centering is equivalent to the formulation we have written down, but I haven't written it out to check?